### PR TITLE
feat: add types to webRequest filter

### LINKED ISSUE #10867: Add types property to onHeaderReceived's filter parameter.
According to [this](https://developer.chrome.com/extensions/webRequest) page, the filter that gets passed to `onHeadersReceived` is a `RequestFilter`, which has a `types` property to allow filtering on resource types. Can this be added to electron's `onHeadersReceived` function, which according to [this](https://github.com/electron/electron/blob/master/docs/api/web-request.md), only supports the `urls` filter?

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
Fixes #10867

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added `types` to `webRequest` filter.

### CODE REVIEW

**Review by erickzhao — APPROVED** [review]
API LGTM

File: docs/api/structures/web-request-filter.md [hunk 1]
```diff
@@ -1,3 +1,4 @@
 # WebRequestFilter Object
 
 * `urls` string[] - Array of [URL patterns](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns) that will be used to filter out the requests that do not match the URL patterns.
+* `types` String[] (optional) - Can be `mainFrame`, `subFrame`, `stylesheet`, `script`, `image`, `font`, `object`, `xhr`, `ping`, `cspReport`, `media` or `webSocket`.
```
Comment (nornagon) [inline docs/api/structures/web-request-filter.md:4]: What does this do? What is the default?

File: shell/browser/api/electron_api_web_request.h [hunk 2]
```diff
@@ -123,20 +123,35 @@ class WebRequest : public gin::Wrappable<WebRequest>, public WebRequestAPI {
...
+    std::set<extensions::WebRequestResourceType> types;
+
+    RequestFilter(std::set<URLPattern>,
+                  std::set<extensions::WebRequestResourceType>);
+    RequestFilter(const RequestFilter&);
+    RequestFilter();
+    ~RequestFilter();
+
+    bool MatchesURL(const GURL& url) const;
+    bool MatchesType(extensions::WebRequestResourceType type) const;
```
Comment (nornagon) [inline shell/browser/api/electron_api_web_request.h:137]: these two could be private?

File: spec/api-web-request-spec.ts [hunk 3]
```diff
@@ -75,8 +75,8 @@ describe('webRequest module', () => {
       await expect(ajax(defaultURL)).to.eventually.be.rejected();
     });
 
-    it('can filter URLs', async () => {
-      const filter = { urls: [defaultURL + 'filter/*'] };
+    it('can filter URLs and types', async () => {
+      const filter: Electron.WebRequestFilter = { urls: [defaultURL + 'filter/*'], types: ['xhr'] };
```
Comment (nornagon) [inline spec/api-web-request-spec.ts:79]: make a separate test please!

**Review by nornagon — CHANGES_REQUESTED** [review]

File: shell/browser/api/electron_api_web_request.h (hunk 2 above)
Comment (miniak) [reply shell/browser/api/electron_api_web_request.h:141]: converted to a proper class

File: spec/api-web-request-spec.ts (hunk 3 above)
Comment (miniak) [reply spec/api-web-request-spec.ts:86]: done

File: docs/api/structures/web-request-filter.md (hunk 1 above)
Comment (miniak) [reply docs/api/structures/web-request-filter.md]: I though it was quite self-explanatory. Default behavior description added.

**Review by zcbenz — APPROVED** [review]
API LGTM

**Review by nornagon — APPROVED** [review]
API LGTM

**Pull request merged**